### PR TITLE
API and tests for budget logic

### PR DESCRIPTION
**Pull request by Andrei-Masilevich**

### CODE REVIEW

File: libraries/chain/dbs_budget.cpp [hunk 1]
```diff
@@ -204,13 +260,62 @@ bool dbs_budget::_check_autoclose(const budget_object& budget)
...
+    dbs_account& account_service = db().obtain_service<dbs_account>();
+
+    const auto& owner = account_service.get_account(budget.owner);
+
+    // withdraw all balance rest asset back to owner
+    //
+    asset repayable = budget.balance;
+    if (repayable.amount > 0)
+    {
+        // check if input budget state != budget in DB
```
Comment (alexplevako) [inline libraries/chain/dbs_budget.cpp]: Does this comment is valid?

File: tests/tests/budget_service_tests.cpp [hunk 2]
```diff
@@ -237,45 +494,144 @@ BOOST_AUTO_TEST_CASE(allocate_cash_next_block)
...
     }
     FC_LOG_AND_RETHROW()
 }
 
-BOOST_AUTO_TEST_CASE(is_const_ref_to_same_memory)
+BOOST_AUTO_TEST_CASE(auto_close_fund_budget_by_deadline)
+{
+    try
+    {
+        db_plugin->debug_update(
```
Comment (alexplevako) [inline tests/tests/budget_service_tests.cpp]: move this code in to the function with a meaningful name. This change will increase readability.

File: tests/tests/budget_service_tests.cpp [hunk 3]
```diff
@@ -237,45 +494,144 @@ BOOST_AUTO_TEST_CASE(allocate_cash_next_block)
...
+            db_plugin->debug_update(
+                [&](database&) {
+                    if (budget_service.get_fund_budget_count() > 0)
+                    {
+                        const budget_object& budget = budget_service.get_fund_budgets()[0];
+                        asset cash = budget_service.allocate_cash(budget);
+                        total_cash += cash;
+                    }
+                },
+                default_skip);
```
Comment (alexplevako) [inline tests/tests/budget_service_tests.cpp]: move this code in to the function with a meaningful name.

File: tests/tests/budget_service_tests.cpp [hunk 4]
```diff
@@ -237,45 +494,144 @@ BOOST_AUTO_TEST_CASE(allocate_cash_next_block)
...
+        asset total_cash(0, SCORUM_SYMBOL);
+
+        for (int ci = 0; ci < BLOCK_LIMIT_DEFAULT; ++ci)
+        {
+            generate_block();
+            generate_block();
+
+            db_plugin->debug_update(
+                [&](database&) {
+                    if (budget_service.get_fund_budget_count() > 0)
```
Comment (alexplevako) [inline tests/tests/budget_service_tests.cpp]: why you a checking budgets count? There should be no logical branches in tests.

File: libraries/chain/dbs_budget.cpp (hunk 1 above)
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_budget.cpp]: Consequence of constant rework of logic. I will fix it

File: tests/CMakeLists.txt [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
 if( GPERFTOOLS_FOUND )
     message( STATUS "Found gperftools; compiling tests with TCMalloc")
     list( APPEND PLATFORM_SPECIFIC_LIBS tcmalloc )
 endif()
 
+file(GLOB HEADERS "common/*.hpp")
+file(GLOB COMMON_SOURCES "common/*.cpp")
 file(GLOB UNIT_TESTS "tests/*.cpp")
-add_executable( chain_test ${UNIT_TESTS} ${COMMON_SOURCES} )
+file(GLOB UNIT_BUDGET_TESTS "tests/budget_check/*.cpp")
```
Comment (abolinbsu) [inline tests/CMakeLists.txt]: Use GLOB_RECURSE

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -1,13 +1,15 @@
...
+
+add_executable( chain_test ${COMMON_SOURCES} ${UNIT_TESTS} ${UNIT_BUDGET_TESTS} ${HEADERS} )
```
Comment (abolinbsu) [inline tests/CMakeLists.txt]: Use target_include_directories

File: libraries/chain/dbs_budget.cpp [hunk 7]
```diff
@@ -51,101 +71,105 @@ dbs_budget::budget_refs_type dbs_budget::get_budgets(const account_name_type& ow
...
+{
+    return db_impl().get<budget_object>(id);
+}
+
+const budget_object& dbs_budget::create_fund_budget(const asset& balance_in_scorum, const time_point_sec& deadline)
 {
     FC_ASSERT(balance_in_scorum.symbol == SCORUM_SYMBOL, "invalid asset type (symbol)");
     FC_ASSERT(balance_in_scorum.amount > 0, "invalid balance_in_scorum");
-    FC_ASSERT(per_block > 0, "invalid per_block");
+    FC_ASSERT(get_fund_budget_count() < SCORUM_LIMIT_FUND_BUDGETS, "can't created more then ${1} fund budgets",
```
Comment (alexplevako) [inline libraries/chain/dbs_budget.cpp]: can't create more than ...

File: libraries/chain/dbs_budget.cpp [hunk 8]
```diff
@@ -51,101 +71,105 @@ dbs_budget::budget_refs_type dbs_budget::get_budgets(const account_name_type& ow
...
+                                               const time_point_sec& deadline,
+                                               const optional<string>& content_permlink)
 {
     FC_ASSERT(owner.name != SCORUM_ROOT_POST_PARENT, "not allowed for ordinary budget");
     FC_ASSERT(balance_in_scorum.symbol == SCORUM_SYMBOL, "invalid asset type (symbol)");
     FC_ASSERT(balance_in_scorum.amount > 0, "invalid balance_in_scorum");
-    FC_ASSERT(per_block > 0, "invalid per_block");
+    FC_ASSERT(owner.balance >= balance_in_scorum, "insufficient funds");
+    FC_ASSERT(get_budget_count(owner.name) < SCORUM_LIMIT_BUDGETS_PER_OWNER,
+              "can't created more then ${1} budgets per owner", ("1", SCORUM_LIMIT_BUDGETS_PER_OWNER));
```
Comment (alexplevako) [inline libraries/chain/dbs_budget.cpp]: can't create ...

File: libraries/protocol/scorum_operations.cpp [hunk 9]
```diff
@@ -336,5 +336,18 @@ void delegate_vesting_shares_operation::validate() const
     FC_ASSERT(is_asset_type(vesting_shares, VESTS_SYMBOL), "Delegation must be VESTS");
     FC_ASSERT(vesting_shares >= asset(0, VESTS_SYMBOL), "Delegation cannot be negative");
 }
+
+void create_budget_operation::validate() const
+{
+    validate_account_name(owner);
+    validate_permlink(content_permlink);
+    FC_ASSERT(is_asset_type(balance, SCORUM_SYMBOL), "Balance must be SCORUM");
```
Comment (alexplevako) [inline libraries/protocol/scorum_operations.cpp]: Not sure, but I thins message it should be SCR not SCORUM

File: libraries/wallet/wallet.cpp [hunk 10]
```diff
@@ -1316,7 +1317,7 @@ pair<public_key_type, string> wallet_api::get_private_key_from_password(const st
  * wallet.
  */
 annotated_signed_transaction wallet_api::create_account_with_keys(const std::string& creator,
-                                                                  const std::string& new_account_name,
```
Comment (alexplevako) [inline libraries/wallet/wallet.cpp]: why have you changed parameter name?

File: libraries/protocol/scorum_operations.cpp (hunk 9 above)
Comment (Andrei-Masilevich) [reply libraries/protocol/scorum_operations.cpp]: Why are we use SCORUM_SYMBOL, VESTS_SYMBOL?
Let 's change to SCR_SYMBOL, SP_SYMBOL

Comment (Andrei-Masilevich) [reply libraries/protocol/scorum_operations.cpp]: And:

#define SCORUM_SYMBOL  (uint64_t(3) | (uint64_t('T') << 8) | (uint64_t('E') << 16) | (uint64_t('S') << 24) | (uint64_t('T') << 32) | (uint64_t('S') << 40)) ///< SCORUM with 3 digits of precision

Why TESTS?

Comment (Andrei-Masilevich) [reply libraries/protocol/scorum_operations.cpp]: There are a lot of that in old code: 

FC_ASSERT(delegator != delegatee, "You cannot delegate VESTS to yourself");
    FC_ASSERT(is_asset_type(reward_scorum, SCORUM_SYMBOL), "Reward Scorum must be SCORUM");
    FC_ASSERT(is_asset_type(reward_vests, VESTS_SYMBOL), "Reward Scorum must be VESTS");
...

Comment (alexplevako) [reply libraries/protocol/scorum_operations.cpp]: for production SCORUM_SYMBOL -> "SCR"
for test net      SCORUM_SYMBOL -> "TESTS"

Comment (alexplevako) [reply libraries/protocol/scorum_operations.cpp]: Variable renaming is not an object of topic.

File: libraries/chain/dbs_budget.cpp [hunk 11]
```diff
@@ -177,6 +207,36 @@ asset dbs_budget::allocate_cash(const budget_object& budget, const optional<time
...
+                                            const time_point_sec& end_date,
+                                            share_type balance_amount)
+{
+    FC_ASSERT(start_date < end_date, "invalid date interval");
+
+    share_type ret(balance_amount);
+
+    fc::microseconds per_block_time = fc::seconds(SCORUM_BLOCK_INTERVAL);
+    fc::microseconds delta = end_date - start_date;
+
```
[the remark this answers is not in the document]
Comment (AlexShkor) [reply libraries/chain/dbs_budget.cpp]: @Andrei-Masilevich I believe that we should change this before merge

File: tests/tests/budget_check/budget_operations.cpp [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    BOOST_REQUIRE_NO_THROW(close_budget_op.validate());
+}
+
+SCORUM_TEST_CASE(close_budget_operation_check_invalid_owner_name)
+{
+    close_budget_op.owner = "";
+
+    BOOST_REQUIRE_THROW(close_budget_op.validate(), fc::assert_exception);
+
+    close_budget_op.owner = "dskf;k30kfl;kvfakg'04i\'j'[IGRIREW40KIA'AKG'K'RK-]]4-4AP[GIRAIORPIGOPRG";
```
[the remark this answers is not in the document]
Comment (Andrei-Masilevich) [reply tests/tests/budget_check/budget_operations.cpp]: Yes

File: libraries/chain/dbs_budget.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_budget.cpp]: fc::seconds(SCORUM_BLOCK_INTERVAL) returns result in microseconds.
Look:  inline microseconds seconds( int64_t s ) { return microseconds( s * 1000000 ); }

File: libraries/chain/dbs_budget.cpp [hunk 13]
```diff
@@ -177,6 +207,36 @@ asset dbs_budget::allocate_cash(const budget_object& budget, const optional<time
     return ret;
 }
 
+share_type dbs_budget::_calculate_per_block(const time_point_sec& start_date,
+                                            const time_point_sec& end_date,
+                                            share_type balance_amount)
+{
+    FC_ASSERT(start_date < end_date, "invalid date interval");
+
+    share_type ret(balance_amount);
```
[the remark this answers is not in the document]
Comment (Andrei-Masilevich) [reply libraries/chain/dbs_budget.cpp]: ret - short name of return_value

**Pull request merged**